src/ucp/wireup/wireup_cm.c

```diff
@@ -358,87 +406,58 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
     if (status != UCS_OK) {
-        goto out_check_err;
+        goto err;
     }
 
     ep->am_lane = key.am_lane;
 
     status = ucp_cm_ep_init_lanes(ep, &tl_bitmap, &dev_index);
     if (status != UCS_OK) {
-        goto out_check_err;
+        goto err;
     }
 
     /* Replay pending requests from the tmp_pending_queue */
     ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
 
-    /* Don't pack the device address to reduce address size, it will be
-     * delivered by uct_cm_listener_conn_request_callback_t in
-     * uct_cm_remote_data_t */
-    status = ucp_address_pack(worker, ep, &tl_bitmap,
-                              ucp_cm_address_pack_flags(worker), NULL,
-                              &ucp_addr_size, &ucp_addr);
-    if (status != UCS_OK) {
-        goto out_check_err;
-    }
-
-    cm_idx = ucp_ep_ext_control(ep)->cm_idx;
-    if (worker->cms[cm_idx].attr.max_conn_priv <
-        ucp_cm_priv_data_length(ucp_addr_size)) {
-        status    = UCS_ERR_BUFFER_TOO_SMALL;
-        log_level = (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE) ?
-                            UCS_LOG_LEVEL_DEBUG :
-                            UCS_LOG_LEVEL_ERROR;
-
-        ucs_log(log_level,
-                "CM private data buffer is too small to pack UCP endpoint"
-                " info, ep %p service data %lu, address length %lu, cm %p"
-                " max_conn_priv %lu",
-                ep, sizeof(ucp_wireup_sockaddr_data_t), ucp_addr_size,
-                worker->cms[cm_idx].cm, worker->cms[cm_idx].attr.max_conn_priv);
-
-        if (ucp_cm_client_try_fallback_cms(ep)) {
-            /* Can fallback to the next CM to retry getting CM initial config to
-             * fit to CM private data */
-            goto out;
-        }
-
-        goto out_check_err;
-    }
-
-    sa_data = ucs_malloc(ucp_cm_priv_data_length(ucp_addr_size),
-                         "client_priv_data");
-    if (sa_data == NULL) {
-        status = UCS_ERR_NO_MEMORY;
-        goto out_check_err;
+    status = ucp_cm_ep_priv_data_pack(ep, &tl_bitmap, dev_index, &priv_data,
+                                      &priv_data_length);
+    if (status == UCS_ERR_BUFFER_TOO_SMALL) {
+        goto try_fallback;
+    } else if (status != UCS_OK) {
+        goto err;
     }
 
-    ucp_cm_priv_data_pack(sa_data, ep, dev_index, ucp_addr, ucp_addr_size);
-
     params.field_mask          = UCT_EP_CONNECT_PARAM_FIELD_PRIVATE_DATA |
                                  UCT_EP_CONNECT_PARAM_FIELD_PRIVATE_DATA_LENGTH;
-    params.private_data        = sa_data;
-    params.private_data_length = ucp_cm_priv_data_length(ucp_addr_size);
+    params.private_data        = priv_data;
+    params.private_data_length = priv_data_length;
     status                     = uct_ep_connect(ucp_ep_get_cm_uct_ep(ep),
                                                 &params);
-    ucs_free(sa_data);
+    ucs_free((void*)priv_data);
```

Comment:
IMO ucp_cm_ep_priv_data_pack should return void*, not const void**

Reply:
do you mean `void **`? because `ucp_cm_ep_priv_data_pack` allocates the buffer according to ucp address size (it's moved to there to unite common code)